common/src/tor.rs

```diff
@@ -0,0 +1,204 @@
+                socks_address_override,
+                forward_address,
+                auth,
+                onion_port,
+                tor_proxy_bypass_addresses,
+                tor_proxy_bypass_for_outbound_tcp,
+            };
+            Ok(transport)
+        } else {
+            let e = format!("Expected a TorHiddenService comms transport, received: {:?}", transport);
+            Err(ExitCodes::ConfigError(e))
+        }
+    }
+
+    /// Run the Tor instance until the shutdown signal is received
+    pub async fn run(self, mut shutdown_signal: ShutdownSignal) -> Result<(), ExitCodes> {
+        info!(target: LOG_TARGET, "Starting Tor");
+
+        let Tor {
+            data_dir,
+            socks_port,
+            control_port,
+            log_level,
+            log_destination,
+            passphrase,
+            ..
+        } = self;
+
+        let mut tor = libtor::Tor::new();
+
+        tor.flag(TorFlag::DataDirectory(data_dir.clone()))
+            .flag(TorFlag::SocksPort(socks_port))
+            .flag(TorFlag::ControlPort(control_port))
+            .flag(TorFlag::Hush())
+            .flag(TorFlag::LogTo(log_level, LogDestination::File(log_destination)));
+
+        if let Some(passphrase) = passphrase {
+            let hashed_pass = hashed_control_password(passphrase)?;
+            tor.flag(TorFlag::HashedControlPassword(hashed_pass));
+        }
+
+        tor.start_background();
+
+        loop {
+            tokio::select! {
+                _ = shutdown_signal.wait() => {
+                    info!(target: LOG_TARGET, "Shutdown signal received, dropping Tor");
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Produce the hashed control password for Tor passphrase authentication
+fn hashed_control_password(passphrase: String) -> Result<String, pgp::errors::Error> {
+    // https://gitweb.torproject.org/tor.git/tree/src/common/crypto_s2k.c?id=7a489a638911012069981702065b952a5809d249#n172
+    // https://gitweb.torproject.org/stem.git/tree/test/integ/process.py#n151
+    let mut rng = rand::thread_rng();
+    let s2k = StringToKey::new_iterated(&mut rng, HashAlgorithm::SHA1, 0x60);
+    let salt = s2k.salt().expect("StringToKey::new_iterated always fills the salt 🧂");
+    let salt_hex = hex::encode_upper(salt);
+    let key = s2k.derive_key(&passphrase, 20)?;
+    let hash = hex::encode_upper(key);
+
+    Ok(format!("16:{}60{}", salt_hex, hash))
+}
+
+/// Attempt to find 2 adjacent available TCP ports in a given range
+fn get_available_ports(mut port_range: Range<u16>) -> Result<(u16, u16), io::Error> {
+    let ports = port_range
+        .find(|port| ports_available(*port))
+        .map(|p| (p, p + 1))
+        .ok_or_else(|| io::Error::new(io::ErrorKind::AddrNotAvailable, "No ports available for Tor."))?;
+
+    Ok(ports)
+}
+
+/// Check if TCP `port` and `port + 1` are available to bind
+fn ports_available(port: u16) -> bool {
```

Comment:
Ports on the tor protocol are not the same as TCP OS-level ports - they are "virtual" ports scoped to the hidden service address

Reply:
It’s for the local control and socks ports passed as arguments to libtor